rust/agents/relayer/src/msg/serial_submitter.rs

```diff
@@ -0,0 +1,232 @@
+    // Contract tracking interchain gas payments for use when deciding whether sufficient funds
+    // have been provided for message forwarding.
+    interchain_gas_paymaster: Option<Arc<CachingInterchainGasPaymaster>>,
+    // Interface to agent rocks DB for e.g. writing delivery status upon completion.
+    db: AbacusDB,
+}
+
+impl SerialSubmitter {
+    pub(crate) fn new(
+        rx: mpsc::Receiver<SubmitMessageArgs>,
+        inbox_contracts: InboxContracts,
+        outbox: Outboxes,
+        interchain_gas_paymaster: Option<Arc<CachingInterchainGasPaymaster>>,
+        db: AbacusDB,
+    ) -> Self {
+        Self {
+            rx,
+            wait_queue: Vec::new(),
+            run_queue: BinaryHeap::new(),
+            inbox_contracts,
+            outbox,
+            interchain_gas_paymaster,
+            db,
+        }
+    }
+
+    pub fn spawn(mut self) -> Instrumented<JoinHandle<Result<()>>> {
+        tokio::spawn(async move { self.work_loop().await })
+            .instrument(info_span!("serial submitter work loop"))
+    }
+
+    #[instrument(skip_all, fields(ibx=self.inbox_contracts.inbox.inbox().chain_name()))]
+    async fn work_loop(&mut self) -> Result<()> {
+        loop {
+            self.tick().await?;
+            tokio::time::sleep(tokio::time::Duration::from_millis(1000)).await;
+            //tokio::task::yield_now().await;
+        }
+    }
+
+    // Tick represents a single round of scheduling wherein we will process each queue and
+    // await at most one message submission.  It is extracted from the main loop to allow for
+    // testing the state of the scheduler at particular points without having to worry about
+    // concurrent access.
+    async fn tick(&mut self) -> Result<()> {
+        // Pull any messages sent by processor over channel.
+        loop {
+            match self.rx.try_recv() {
+                Ok(msg) => {
+                    self.wait_queue.push(msg);
+                }
+                Err(TryRecvError::Empty) => {
+                    break;
+                }
+                _ => {
+                    bail!("disconnected rcvq or fatal err");
+                }
+            }
+        }
+
+        // TODO(webbhorn): Scan verification queue, dropping messages that have been confirmed
+        // delivered by the inbox indexer observing it.  For any still-unverified messages that
+        // have been in the verification queue for > threshold_time, move them back to the wait
+        // queue for further processing.
+
+        // Promote any newly-ready messages from the wait queue to the run queue.
+        for msg in &self.wait_queue {
+            // TODO(webbhorn): Check if already delivered to inbox, e.g. by another relay. In
+            // that case, drop from wait queue.
+            // TODO(webbhorn): Check against interchain gas paymaster.  If now enough payment,
+            // promote to run queue.
+            info!(msg.leaf_index, "-> runq");
```

Comment:
maybe `debug`? and a more friendly log message may be helpful

Reply:
I've decided to drop this log since it's clear from the code that only one possibility for execution exists.